src/MonitorProfiler/MainProfiler/ExceptionTracker.cpp

```
@@ -0,0 +1,392 @@
+
+    Current pitfalls:
+    - Algorithm doesn't account for exceptions thrown within exception filters. These will be seen as a new set of
+      callback invocations starting with throwing, searching, and unwinding, however, the callstack does not unwind
+      out of the exception filter and the original exception processing is resumed.
+    - Algorithm doesn't account for exceptions thrown within finally blocks. Presumably, these will be seen as new set
+      of callback invocations starting with throwing, searching, and unwinding. The original exception processing is
+      superceded by the new set of callbacks for the new exception.
+*/
+
+
+ExceptionTracker::ExceptionTracker(
+    const shared_ptr<ILogger>& logger,
+    const shared_ptr<ThreadDataManager> threadDataManager,
+    ICorProfilerInfo2* corProfilerInfo)
+{
+    _corProfilerInfo = corProfilerInfo;
+    _logger = logger;
+    _threadDataManager = threadDataManager;
+}
+
+void ExceptionTracker::AddProfilerEventMask(DWORD& eventsLow)
+{
+    eventsLow |= COR_PRF_MONITOR::COR_PRF_ENABLE_STACK_SNAPSHOT;
+}
+
+HRESULT ExceptionTracker::ExceptionThrown(ThreadID threadId, ObjectID objectId)
+{
+    // CAUTION: Do not store the exception ObjectID. It is not guaranteed to be correct
+    // outside of the ExceptionThrown callback without updating it using GC callbacks.
+
+    HRESULT hr = S_OK;
+
+    IfFailLogRet(_threadDataManager->SetHasException(threadId));
+
+    // Exception throwing is common; don't pay to calculate method name if it won't be logged.
+    if (_logger->IsEnabled(LogLevel::Debug))
+    {
+        FunctionID functionId = 0;
+        hr = _corProfilerInfo->DoStackSnapshot(
+            threadId,
+            LogExceptionThrownFrameCallback,
+            COR_PRF_SNAPSHOT_INFO::COR_PRF_SNAPSHOT_DEFAULT,
+            this,
+            nullptr,
+            0);
+
+        if (FAILED(hr) && hr != CORPROF_E_STACKSNAPSHOT_ABORTED)
```

If `hr` is `CORPROF_E_STACKSNAPSHOT_ABORTED`, we currently just silently return `S_OK` - without knowing the full context of the `DoStackSnapshot` method, would it make sense to log anything when this situation occurs?

---

No, we intentionally ending the callback by returning `S_FALSE`, which DoStackSnapshot will return `CORPROF_E_STACKSNAPSHOT_ABORTED`, which is intended.